README.md
# Lob asset management

A solution created to generate investments income tax release 

--------------
## Usage
### Read XLSX movements

Read a XLSX document and turn in EDN entities:

 - Asset
 - Transaction
 - Portfolio

#### Asset

Register all unique `:asset/ticket` from movements and use `resources/asset_fixed_info.edn` to complement information

```Clojure 
:asset/ticket 
:asset/tax-number
:asset/id
:asset.market-price/price
:asset.market-price/price-date
:asset/category
:asset.market-price/updated-at
:asset.market-price/historic 
:asset/name
:asset/type
```

#### Transactions

Financial movements, convert all movements from XLSX documents to EDN

```Clojure 
:transaction/operation-type
:transaction/operation-total
:transaction/currency
:transaction/average-price
:transaction.asset/ticket
:transaction/id
:transaction/processed-at
:transaction/quantity
:transaction/exchange
:transaction/created-at
```

##### Operation Types
```Clojure
:sell ;=> (and (= type"Debito")
      ;        (or (= movement-type "Transferência - Liquidação")
      ;            (= movement-type "COMPRA / VENDA")))
:buy  ;=> (and (= type "Credito")
      ;        (or (= movement-type "Transferência - Liquidação")
      ;            (= movement-type "COMPRA / VENDA")))
:desdobro ; Multiply the asset quantity
:grupamento ; Group the asset quantity
:waste ;Fraction of the asset was sold
:dividend
:JCP
:income ;Dividend from Reit assets
:bonificaçãoemativos ;Bonus in asset, increase the asset quantity
:incorporation ;Come from incorporation_movement file in source folder
;;;;IGNORED
:fraçãoemativos ;IGNORED [No considerable values]
:compra ;IGNORED [Government Bound]
:transferência  ;IGNORED
:solicitaçãodesubscrição ;IGNORED
:cessãodedireitos-solicitada  ;IGNORED
:direitodesubscrição  ;IGNORED
:cessãodedireitos ;IGNORED
:direitosdesubscrição-excercído ;IGNORED
:direitosdesubscrição-nãoexercido ;IGNORED
:recibodesubscrição ;IGNORED
:compraporliquides  ;IGNORED
:vencimento ;IGNORED
:resgate  ;IGNORED
:incorporação ;IGNORED
:atualização  ;IGNORED
```

##### Incorporation event

Incorporation events needs to be registered in the file `resources/incorporation_movement` fallowing this format :

```Clojure
[{:transaction-date "21/06/2022" ;Date of the incorporation
  :unit-price       "0,00"       ;Will be all position value when processed
  :quantity         0            ;Will be all position value when processed
  :exchange         "INTER DISTRIBUIDORA DE TITULOS E VALORES MOBILIARIOS LTDA"
  :product          "BIDI11 - BANCO INTER S/A" ;The acquired company ticket
  :operation-total  "0,00"       ;Will be all position value when processed
  :currency         "BRL"        ;BRL or USD
  :movement-type    "Incorporation" ;Turn in transaction/operation-type 
  :incorporated-by  "INBR32"     ;The new asset owner
  :factor           "/2"         ;operator can be "*" or "/" and fallowing by the denominator 
  }]
``` 

#### Portfolio

The consolidation of transactions grouped by `:transaction.asset/ticket`

```Clojure 
:portfolio/transaction-ids
:portfolio/quantity
:portfolio/percentage
:portfolio/sell-profit
:portfolio/total-cost
:portfolio/dividend
:portfolio/total-last-value
:portfolio/category
:portfolio/exchanges
:portfolio/average-price
:portfolio/ticket
:portfolio.profit-loss/percentage 
:portfolio.profit-loss/value
```
--------------
### Portfolio Balance [TODO]

1. Throw Asset type (StockBR StockEUA Crypto) by configuration
   1. Current situation gain/loss focusing in the lowest profit (Configuration first)
2. Throw Categories allocation
   1. To compare current allocation percentage with configured allocation
3. Throw Category assets balance allocation 
   1. Maintain close the asset distribution
   2. Verify the current gain/loss focusing in assets with loss
   3. Identify the best asset (Classify assets) / Determined asset configured allocation (throw category)

**Rules**

- Different configuration throw types (StockBR - StockEUA - Crypto)
  - StockBR we use quantity and StockEUA/Crypto we use total money to allocate
  - StockBR => Buy at least one asset quantity
  - StockEUA/Crypto => Distribute the amount based in the category allocation and asset priority
- Configure asset allowed to be purchased
- Configure asset priority
  - options :
    - [ ] Fixed Percentage (ex 50)
    - [ ] Classification range number (ex 1~5)
    - [ ] Recent or/and total decrease in price have most priority
    - [ ] Composition of recent/total decrease in value plus classification

**Flow StockBR**
1. Determine how much money will be allocated
2. Buy at least one position for each allowed asset
3. Current category allocation with the previews step X category configuration
4. Divide the rest of the money following the reminder needs to archive the configuration by category
5. Follow the asset priority configuration to choose the asset

**Flow Stock EUA**
1. Determine how much money will be allocated
2. Check the current category allocation x configuration
3. Divide the total between category to archive the goal
4. Divide the total in category by asset fallowing asset priority configuration

**Flow for Crypto**
1. Determine how much money will be allocated
2. Divide the total between asset priority

--------------
## How allow Telegram Bot

1. [Obtain Your Bot Token](https://core.telegram.org/bots/tutorial#obtain-your-bot-token)
2. Set the key as `telegram-bot-key` environment variable

--------------
## TODO

1. Provide where asset need to be purchase and how much `high
   1. [X] v1 Get allocation needs following configuration 
   2. [ ] v2 Consider extra budget
   3. [ ] v3 Consider recent asset performance
   4. [ ] vX Simulator
2. Store data `Medium`
   1. [ ] AWS
      1. ~~Thinking about the item 4. maybe we can use S3 only for the in-data files~~
   2. [ ] Firebase (Doing)
      1. [x] Find firebase library (alekcz/fire)
      2. [x] Create firebase project 
      3. [ ] Fix connection only works in terminal (lein run)
      4. [ ] Create write/read documents
   3. [ ] Do nothing
      1. Solution works for read/write entities in terminal
      2. ~~The problem is only **maybe** in the in-data files~~
   4. XTDB 
3. [Http_in] Find another option for Alpha API / Web Scraping `Medium`
   1. [X] Coingecko for crypto 
   2. [ ] Web Scraping for Stock information
4. Consider subscription events `Low`
5. Get asset information `Very low`
6. Decision between stock APIs
   1. [ ] Validate if the historic price is updated
   2. [ ] Get more often with real time prices API
   3. [ ] When error in one get using another
      1. Create a new ident for error (E.g :asset/recent-alpha-api-error? true)
      2. If true => Use another
8. Fix the telegram scheduled messages [CRON]
9. Make the poling async and different by proposal (Get price / Get news / TelegramBot)


## FIXME
1. Consider ALZR subscription

--------------
## License

Copyright © 2023 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### 2023-03-25
- Project created
- Read xmlx files
- Data transformation for
    - Asset
    - Transaction

### 2023-03-31
- Read and write edn files
- Backup file before write new file

### 2023-04-01
- Read the b3 movement and write edn file for
    - asset
    - transaction
- Filter asset to be unique records
- Concat stored file data with read from b3 movement
- Remove duplicated when concat stored with b3 movement
- New folders
    - controller
    - logic

### 2023-04-18
- Validate duplicated transactions
- DR 001 - Unique by ID and don't treat update/changes in the comparative

### 2023-04-21
- Transaction to portfolio average price calculation
- Concluded transaction to portfolio adepter

### 2023-04-25
- Add portfolio file (read/write)
- Add schema function validation
- ReFactory io functions
- Create delete file function

### 2023-04-26
- Read only new files function
- Add support to upsert a simple map


### 2023-04-28
- Started Read B3 prices
- Create a secrets file
- Create relevant namespace for global config and secrets

### 2023-04-30
- Add percentage by asset in portfolio
- Add category field in portfolio
- Add percentage by category

### 2023-05-02
- Register profit by asset (Dividend, JCP...)

### 2023-05-03
- Create Market controller to get asset prices

### 2023-05-04
- Include get less updated asset
- Update file with last market price
- Fix get only new releases process

### 2023-05-06
### 2023-05-07
### 2023-05-08
- Fix process file flow
- Fix filter for assets and transactions
- Create Polling function

### 2023-05-10
- Create a file with fixed asset information
- Started creation of irpf release
- Add get asset overview endpoint [don't have b3 asset info]

### 2023-05-11
- Add fixed asset info in asset
- Add allow alpha api asset in fixed info file
- Change way to get asset category to fixed asset info
- Create function to format br tax document

### 2023-05-12
- Fix release description
- Read all data from B3 with success with some fixes in asset and transaction adapters

### 2023-05-18
- Read/Process stock info
- Created config file for xlsx info
- Add log library
- Add price historic in asset
- Set poller configuration

### 2023-05-22
- Read crypto info
- Store stock info
- Update release group and code
- Read all releases by configuration

### 2023-05-23
- Generate Income Tax release file
- Get EUA stock price
- Fix market historic price format
- Change println for log message

### 2023-05-24
- Add get and register USD price

### 2023-05-31
- Chance portfolio allocation to use current asset price

### 2023-06-02
- Get crypto market price

### 2023-07-05
- Count Alpha API calls
- Telegram message with API calls

### 2023-07-07
- Add Coingecko API for real time crypto prices

### 2023-07-12
- Process global incorporation event

### 2023-07-13
- Refactor portfolio process
  + Portfolio Controller
  + Portfolio Database

### 2023-07-14
- Consider incorporation in portfolio process
- Fix db.portfolio/upsert!
- Backup cleanup

### 2023-07-18
- Throw and treat alpha api limit exeption

### 2023-07-20
- Add transaction DB
- Add asset DB
- Changed all read on [transaction, asset, portfolio] direct from io.f-in to db

### 2023-07-24
- Add telegram allowed commands message

### 2023-07-26
- Change Alpha API function from TIME_SERIES_DAILY_ADJUSTED to TIME_SERIES_DAILY due changed to premium plan

### 2023-08-05
- Add bot save message
- Add bot message from validation

### 2023-08-15
- Move telegram key to environment variable

### 2023-09-19
- Remove clj-time library
- Change asset news database to use XTDB

### 2023-09-23
- Asset news bot message

### 2023-10-21
- Price change notification
- Telegram stop command

### 2023-12-15
- Create new Poller schedule based
- Move corer poller to scheduled tasks

## [0.1.1] - 2023-03-25
### Changed

### Removed

### Fixed

## 0.1.0 - 2023-03-25
### Added

[Unreleased]: https://github.com/your-name/lob-asset-management/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/lob-asset-management/compare/0.1.0...0.1.1


src/lob_asset_management/core.clj
(ns lob-asset-management.core
  (:require [clojure.tools.cli :as t.cli]
            [clojure.tools.logging :as log]
            [java-time.api :as t]
            [lob-asset-management.aux.time :as aux.t]
            [lob-asset-management.aux.xtdb :refer [db-node]]
            [lob-asset-management.controller.forex :as c.f]
            [lob-asset-management.controller.market :as c.m]
            [lob-asset-management.controller.process-file :as c.p-f]
            [lob-asset-management.controller.portfolio :as c.p]
            [lob-asset-management.controller.release :as c.r]
            [lob-asset-management.controller.scheduler :as c.s]
            [lob-asset-management.controller.telegram-bot :as t.bot :refer [bot]]
            [lob-asset-management.db.asset :as db.a]
            [lob-asset-management.db.forex :as db.f]
            [lob-asset-management.db.portfolio :as db.p]
            [lob-asset-management.io.file-in :as io.f-in]
            [lob-asset-management.relevant :refer [config config-prod]]
            [mount.core :as mount]))

(defn start [env]
  "Default env = :dev

  v2 => mount functions to be used by environment"
  (mount/start #'lob-asset-management.relevant/config
               #'lob-asset-management.relevant/alpha-key
               #'lob-asset-management.relevant/telegram-key
               #'lob-asset-management.relevant/telegram-personal-chat
               #'lob-asset-management.controller.telegram-bot/bot)
  (when (= :prod env)
    (mount/start #'lob-asset-management.aux.xtdb/db-node)
    (mount/start-with {#'lob-asset-management.relevant/config config-prod})))

(defn stop []
  (mount/stop #'lob-asset-management.relevant/config
              #'lob-asset-management.relevant/alpha-key
              #'lob-asset-management.relevant/telegram-key
              #'lob-asset-management.relevant/telegram-personal-chat
              #'lob-asset-management.controller.telegram-bot/bot
              #'lob-asset-management.aux.xtdb/db-node))

;(start :dev)                                                     ;for develop purpose
;(stop)

(def cli-options
  [["-y" "--year Year" "Year of the release"
    :default 2022
    :parse-fn #(Integer/parseInt %)
    :validate [#(< 2021 % 2024) "Must be a number between 2021 and 2023"]]
   ["-d" "--[no-]daemon" "Daemonize the process" :default true]
   ["-h" "--help"]])

(defn usage [options-summary]
  (->> ["This is my program. There are many like it, but this one is mine."
        ""
        "Usage: program-name [options] action"
        ""
        "Options:"
        options-summary
        ""
        "Actions:"
        "  start    Start the get market price and update portfolio"
        "  read     Read the movements files"
        "  release  [send the year as -y parameter] Generate a new release"
        "  telegram [send the message as -m parameter] Send a message to telegram"
        ""
        "Please refer to the manual page for more information."]
       (clojure.string/join \newline)))

(defn error-msg [errors]
  (str "The following errors occurred while parsing your command:\n\n"
       (clojure.string/join \newline errors)))

(defn validate-args
  "Validate command line arguments. Either return a map indicating the program
  should exit (with an error message, and optional ok status), or a map
  indicating the action the program should take and the options provided."
  [args]
  (let [{:keys [options arguments errors summary]} (t.cli/parse-opts args cli-options)]
    (cond
      (:help options) ; help => exit OK with usage summary
      {:exit-message (usage summary) :ok? true}
      errors ; errors => exit with description of errors
      {:exit-message (error-msg errors)}
      ;; custom validation on arguments
      (and (= 1 (count arguments))
           (#{"start" "read" "release" "telegram"} (first arguments)))
      {:action (first arguments) :options options}
      :else ; failed custom validation => exit with usage summary
      {:exit-message (usage summary)})))

(defn exit [status msg]
  (println msg)
  (System/exit status))

(defn get-market-info
  [forex-usd stock-window current-time]
  (let [assets (db.a/get-all)
        portfolio (db.p/get-all)
        current-hour (.getHour current-time)
        day-of-week (aux.t/day-of-week current-time)]
    (c.m/reset-retry-attempts assets)
    (if (contains? stock-window current-hour)
      (when (c.m/update-asset-market-price assets day-of-week)
        (c.p/update-portfolio-representation portfolio forex-usd))
      (when (c.m/update-crypto-market-price assets)
        (c.p/update-portfolio-representation portfolio forex-usd)))))

(defn start-processing
  [stock-window interval]
  (let [forex-usd (db.f/get-all)
        update-target-hour 3
        current-time (t/local-date-time)]
    (t.bot/check-telegram-messages interval current-time)
    (c.p-f/backup-cleanup :asset)                           ;FIXME: Didn't work in the shell execution
    (if (c.f/less-updated-than-target forex-usd update-target-hour)
      (c.f/update-usd-price forex-usd update-target-hour)
      (get-market-info forex-usd stock-window current-time))))

(defn poller [f-name f interval window]
  (let [run-forest-run (atom true)]
    (future
      (try
        (while @run-forest-run
          ;(log/info "[Poller running" (str (t/local-date-time)) "]")
          (if (contains? window (.getHour (t/local-date-time)))
            (f)
            (log/info "[Poller " (str f-name "-" (t/local-date-time)) "] Out of the configured window hour " window))
          (log/info "[Poller next " (str f-name "-" (t/plus (t/local-date-time)
                                                            (t/millis interval))) "]")
          (Thread/sleep interval))
        (catch Exception e
          (println "Error in " f-name " poller: " e))))
    (fn [] (reset! run-forest-run false))))

(defn -main [& args]
  (start :prod)
  (let [{:keys [action options exit-message ok?]} (validate-args args)]
    (if exit-message
      (exit (if ok? 0 1) exit-message)
      (case action
        "start" (let [interval 3000
                      stop-loop (c.s/poller interval)
                      #_(poller "Main"
                              #(start-processing #{11 12 13 14 15 16 17 18 19 20 21 22 23} interval)
                              13000
                              #{7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 00 01})]
                  (println "Press enter to stop...")
                  (read-line)
                  (future-cancel (stop-loop))
                  @(stop-loop))
        "read"    (c.p-f/process-folders)
        "release" (c.r/irpf-release (:year options))))))

(comment
  (schema.core/set-fn-validation! true)

  (c.p-f/delete-all-files)
  (c.p-f/process-folders)
  ;=========================================

  (clojure.pprint/print-table
    [:portfolio/ticket :portfolio/quantity]
    (->> (db.p/get-all)
         (filter #(contains? #{:nu :inter} (first (:portfolio/exchanges %))))
         ;(filter #(contains? #{:sproutfy} (first (:portfolio/exchanges %))))
         ;(filter #(= :crypto (:portfolio/category %)))
         ;(filter #(or (contains? (:portfolio/exchanges %) :nu)
         ;             (contains? (:portfolio/exchanges %) :inter)))
         (sort-by :portfolio/ticket)))

  (filter #(= :SMTO3 (:portfolio/ticket %)) (db.p/get-all))

  (clojure.pprint/print-table
    [:transaction/id :transaction/created-at :transaction/operation-type :transaction/quantity :transaction/average-price]
    (->> (lob-asset-management.db.transaction/get-all)
         ;(filter #(= :fraçãoemativos (:transaction/operation-type %)))
         (filter #(or (= :SULA11 (:transaction.asset/ticket %))))
         (remove #(contains?
                    #{:sell :JCP :income :dividend :bonificaçãoemativos
                      :fraçãoemativos :transferência :waste :incorporation}
                    (:transaction/operation-type %)))
         #_(filter #(or (= :SQIA3 (:transaction.asset/ticket %))
                      (= :EVTC31 (:transaction.asset/ticket %))))
         ;(filter #(contains? #{:sproutfy} (:transaction/exchange %)))
         ;(sort-by :transaction/exchange)
         ;(sort-by :transaction.asset/ticket)
         (sort-by :transaction/created-at)
         ))

  (->> (lob-asset-management.db.transaction/get-all)
       ;(filter #(= :fraçãoemativos (:transaction/operation-type %)))
       (filter #(or (= :SULA11 (:transaction.asset/ticket %))))
       (remove #(contains?
                  #{:sell :JCP :income :dividend :bonificaçãoemativos
                    :fraçãoemativos :transferência :waste :incorporation}
                  (:transaction/operation-type %)))
       #_(filter #(or (= :SQIA3 (:transaction.asset/ticket %))
                      (= :EVTC31 (:transaction.asset/ticket %))))
       ;(filter #(contains? #{:sproutfy} (:transaction/exchange %)))
       ;(sort-by :transaction/exchange)
       ;(sort-by :transaction.asset/ticket)
       (sort-by :transaction/created-at)
       (reduce #(+ %1 (:transaction/quantity %2)) 0))
  ;=========================================
  (def in (lob-asset-management.relevant/incorporation-movements))

  (c.p-f/process-movement in)

  (def m (io.f-in/get-file-by-entity :metric))
  (def tm (filter #(and (= "https://www.alphavantage.co/query" (-> % :endpoint :url))
                        (= :2023-07-17 (aux.t/clj-date->date-keyword (:when %))))
                  (:metric/api-call m)))


  (def c (atom 0))
  (map #(let [f (-> % :endpoint :params :query-params :function)
              s (-> % :endpoint :params :query-params :symbol)]
          (swap! c inc)
          {:x (str "[" @c "]" f " - " (or s ""))}) tm)

  (c.m/update-crypto-market-price)
  (c.p/update-portfolio-representation (db.p/get-all) (db.f/get-all)))
